### DISCUSSION TITLE
packeto buildpacks are still using go 1.20 which is unsupported / EOL triggering security alerts

### DISCUSSION THREAD

**Discussion by candrews** [paketo-buildpacks/feedback#283] [General]
I noticed that many Paketo Buildpacks projects are still using go 1.20 which is EOL / unsupported according to https://go.dev/doc/devel/release

I discovered this issue because security scanners are now reporting vulnerabilities against Paketo Buildpacks and the images they build. For example, CVE-2023-45288 is reported frequently and being a high severity vulnerability it sets off some loud alarms. I'm confident that this vulnerability and the others reported are not exploitable, but their presence is still troublesome for it still requires effort to suppress the findings and justify the reasoning.

Can all usages of go 1.20 please be upgraded to at least go 1.21, ideally 1.22?

I see that Paketo used to have an automatic process for doing go upgrades, as I see many PRs like https://github.com/paketo-buildpacks/native-image/pull/251 by @paketo-bot. Perhaps that automation can be revived?

**Comment by sophiewigmore** [comment]
Thanks for the ping @candrews. We should definitely bump the Go version, and try to automate doing this in the future

**Reply by dmikusa** [reply]
I'm going to look at this soon for Java buildpacks. I believe we had some issues when one of our buildpacks accidentally got updated to 1.21 (something with module changes). I'll post any issues I encounter here.

**Reply by dmikusa** [reply]
From the Go 1.21 release notes:

> To improve forwards compatibility, Go 1.21 now reads the go line in a go.work or go.mod file as a strict minimum requirement: go 1.21.0 means that the workspace or module cannot be used with Go 1.20 or with Go 1.21rc1. This allows projects that depend on fixes made in later versions of Go to ensure that they are not used with earlier versions. It also gives better error reporting for projects that make use of new Go features: when the problem is that a newer Go version is needed, that problem is reported clearly, instead of attempting to build the code and printing errors about unresolved imports or syntax errors.

FYI only. 

I think this is what we hit before. If I remember correctly, a `go.mod` version got bumped but not the runtime, so it was causing problems. Seems like this really shouldn't be an issue so long as the runtimes being installed in CI get bumped in conjunction with the versions set in `go.mod` files.

**Reply by sophiewigmore** [reply]
As long as all of our workflows use `go-version: 'stable'` in their Go setup step, should that do the trick @dmikusa?

**Reply by dmikusa** [reply]
I'm not 100% sure @sophiewigmore. I've not used `stable`, we use a specific Go version, but that seems kind of nice. A rolling, up-to-date version. 

The other thing we update is the `go.mod` file to have `go 1.22` (or whatever version we're on at the moment). I'm not sure if both are necessary or if it is sufficient to just bump the SDK that's used to build the binary. I'd be curious to know if anyone can explain that more or link to something that does.

**Reply by ForestEckhardt** [reply]
My understanding is that the version specified in the `go.mod` file is the minimum viable compilation version to ensure compatibility and it will compile with the latest compiler.

**Reply by robdimsdale** [reply]
There are two versions in `go.mod`

`go 1.x` - the minimum version of go (language specification) required to compile the module and its dependencies (e.g. libraries using a syntax only available in newer versions of go).

`toolchain go1.x.y` - the exact version of the go (toolchain ). If it is greater than or equal to the `go` version then this version is used (and downloaded if necessary). If the `toolchain` declaration is present, it must be greater than or equal to the `go` version - otherwise `go mod tidy` will remove the line.

One wrinkle in this, is that you need a minimum version of `go 1.21` in order to use the `toolchain` command, even if your code (and your dependencies) could be compiled with a lower version of go.

All of that being said, I would prefer to see `toolchain go1.x.y` in `go.mod` rather than relying on versions in CI. This is primarily because it is easier to reason about which SHA in a git repository would be built with which version. I accept that for GitHub CI this is less of an issue if the workflows have a specific version for `go-version`, but if you have `go-version: stable`, or switch to a CI system that is decoupled from the buildpack repository then this becomes non-optional.

**Reply by dmikusa** [reply]
Thanks @robdimsdale & @ForestEckhardt for explaining that. Very helpful.

I think one of the issues we saw, just prior to upgrading, was that we must have had dependency libraries that updated to go 1.21 and when we updated to them, despite not having major version changes in the dependencies, it broke our builds. One of the go commands, probably `go mod tidy` based on what you're saying, was bumping that minimum version in our `go.mod` file and then our older Go 1.20 was failing to build, cause it's older. It seemed odd to me that the tooling would change our `go.mod` file, but again, your explanation makes what it was doing make sense.

There's certainly something appealing about setting `go-version: stable` and never really having to worry about any of this again. The 6 month cycle of updating Go versions has always been a bit of a pain for us. I do like having the specific Go version somewhere though, like you said for historical reasons.

It would make sense to me if we did this the same across all of Paketo as well. Would you be interested in picking a way to do it and proposing a path forward in an RFC? i.e. put X into GHA workflows, put Y into go.mod, then we get consistent results.

**Reply by robdimsdale** [reply]
yes, I don't fully understand all the situations in which `go mod tidy` will edit the `go.mod` file. It seems to be a function of:
- the version of go toolchain that is installed (i.e. what is actually running `go mod tidy`)
- what go version/toolchain version is already in `go.mod`
- what go version/toolchain version is in the `go.mod` files across the transitive dependencies
- ???

But, I have seen that if we have `go 1.XX` and `toolchain go1.X.Y` already in the `go.mod`, and we're using `go 1.21.x` or later when we run the `go mod tidy` command, then usually the `go.mod` file is unedited.

### Proposal

So, my proposal would be to use `stable` in CI, and create new automation to update (and hence pin) the `toolchain` version in the `go.mod` files. This could be run periodically and/or manually, depending on buildpack authors' preferences. I'd vote for periodically (e.g. weekly) as I don't monitor new go releases - especially not patch releases.

I'd be open to running this automatoin more frequently (e.g. daily), but until we release buildpacks more frequently than weekly I think looking for new versions of `go` daily is a waste of CI resources.

---

Separately from the above automation, at some point we'd bump a dependency to a version that would have a later `go` version in its `go.mod`, and I believe that process would cause the buildpack to bump the `go` version in `go.mod` to the same version as the dependency.

Concretely:

Given:
- the buildpack currently has `go 1.21` in the `go.mod`
- the buildpack depends on `library-a` at version `v1.2.3`
- `library-a` at `v1.2.3` has `go 1.21` in its `go.mod`

When:
- if `library-a` version `v1.3.0` has `go1.22` in its `go.mod`
- buildpack bumps `library-a` from `v1.2.3` to `v1.3.0`

Then:
- buildpack's `go.mod` will now contain `go 1.22` in the `go.mod`

I think that this process would happen as part of dependency bumps, for example during Dependabot updates. I don't think we need to do anything to bump the `go` version in `go.mod`. Just the `toolchain` version.

**Reply by sophiewigmore** [reply]
@robdimsdale thank you SO much for the in-depth explanation. Since the toolchain changes in 1.21, I've been quite confused on how this behaviour is meant to work, this helped a lot. I like your proposal to keep go updated as well 👍

**Reply by dmikusa** [reply]
Sounds good to me.

**Reply by robdimsdale** [reply]
PR to add the above here: https://github.com/paketo-buildpacks/github-config/pull/928

**Reply by robdimsdale** [reply]
PR has been merged - we should see `toolchain` getting written when the workflows run, starting next monday.

**Reply by dmikusa** [reply]
Is there someone that can follow through and make sure we end up releasing buildpacks after this change has been applied? 

I know for a fact that web-servers & nginx (the culprit) are using Go 1.22.2 which has a flagged CVE so images from those buildpacks are generating images flagged with CVEs. We probably don't need to release all the buildpacks, but any buildpack that inserts an exec.d helper that's a Go binary should.  

For example:

```
layers/paketo-buildpacks_nginx/nginx/exec.d/0-configure (gobinary)
==================================================================
Total: 1 (HIGH: 1, CRITICAL: 0)

┌─────────┬────────────────┬──────────┬────────┬───────────────────┬───────────────┬────────────────────────────────────────────────────────────┐
│ Library │ Vulnerability  │ Severity │ Status │ Installed Version │ Fixed Version │                           Title                            │
├─────────┼────────────────┼──────────┼────────┼───────────────────┼───────────────┼────────────────────────────────────────────────────────────┤
│ stdlib  │ CVE-2024-2[47](.../actions/runs/9356761067/job/25755033257#step:5:48)88 │ HIGH     │ fixed  │ 1.22.2            │ 1.22.3        │ golang: net: malformed DNS message can cause infinite loop │
│         │                │          │        │                   │               │ https://avd.aquasec.com/nvd/cve-2024-24788                 │
└─────────┴────────────────┴──────────┴────────┴───────────────────┴───────────────┴────────────────────────────────────────────────────────────┘
```

**Comment by candrews** [comment]
Would it be helpful for me to submit PRs to the various projects doing this work?

I'm eager to get this done as it's causing a lot of headache and paperwork for my Powers That Be.

**Reply by dmikusa** [reply]
I work on about half of the buildpacks, and we're rolling out a bump to 1.22 shortly. It takes a little time to roll through all the CI & for releases to be cut.

What buildpacks are you specifically looking for? Those I don't work on, I can inquire more about.

**Reply by candrews** [reply]
The ones I'm specifically using are:
* paketo-buildpacks/ca-certificates
* paketo-buildpacks/bellsoft-liberica
* paketo-buildpacks/syft
* paketo-buildpacks/executable-jar
* paketo-buildpacks/dist-zip
* paketo-buildpacks/spring-boot

**Reply by dmikusa** [reply]
OK. I'm hoping to have all of those updated and released this week.

**Reply by anthonydahanne** [reply]
hello @dmikusa , isnt it all done (at least for Java buildpacks?)
Thanks

**Reply by dmikusa** [reply]
Yes, they were all updated the week of May 22nd.

**Reply by candrews** [reply]
I checked `paketo-buildpacks/ca-certificates` and `paketo-buildpacks/bellsoft-liberica`. Both are using 1.22.3 which is great, addressing the request of this issue.

However, both are using go 1.22.3 which does have some findings and is not the latest version of go.

Can the buildpacks use go 1.22.4, and keep updating to the latest versions of go as they become available?

**Reply by dmikusa** [reply]
Ugh, more security issues. 1.22.3 was the most recent when we released last time. We'll do another round. Hopefully the Go CVEs slow down soon :)

**Reply by candrews** [reply]
Do you have an issue you're using to track the go 1.22.4 upgrade?

I'm getting a lot of noise due to CVE-2024-24790 being reported in the buildpacks so I'm really looking forward to this update.

Perhaps you could use dependabot, renovate, or some other tool to keep updating the go version going forward, eliminating the need for manual work?

**Reply by dmikusa** [reply]
It isn't updating it that's the problem. Go gets updated automatically for patch releases. It's that buildpacks need to be re-released, that is more work and not something we can commit to for *every* CVE reported.

> We aim to release stack updates that fix High and Critical CVEs within 48 hours of the patch release. For stack updates fixing Low and Medium CVEs, we aim to release within two weeks.

https://paketo.io/docs/concepts/stacks/#when-are-paketo-stacks-updated

We follow the same policy for buildpacks and when we'll cut new releases for CVEs. CVEs that do not impact buildpacks would also be in the two-week category, but please take note that is best-effort, not a guarantee.

**Reply by candrews** [reply]
CVE-2024-24790 is a Critical CVE - so shouldn't it be addressed within 48 hours of the patch release?

**Reply by dmikusa** [reply]
How would that impact buildpacks? That's the thing with buildpacks, we use a very small part of the Go runtime, most CVEs are not even related to methods we're using. Even if we did use one, because of what buildpacks are doing it would be very hard to exploit in a meaningful way.

The primary use case we watch out for are with exec.d processes. These get put into your application container, so they could impact runtime which is the biggest concern. Most/all of these that we install are very simple. They look at env variables and set env variables. They're not doing anything network-related (I don't know of any that do). So there shouldn't be anything at runtime invoking Go runtime network calls. That stuff happens at build time, but there are way easier ways to do whatever you want at build time without a CVE being required, so again they don't meaningfully impact buildpacks.

tl;dr. We look at Go CVEs, but 99% of them are not going to impact buildpacks in a way that would be HIGH/CRITICAL, so they fall into the two-weeks window.
